File: Algs4/src/1.Basic/BinarySearch.java
import edu.princeton.cs.algs4.In;
import edu.princeton.cs.algs4.StdIn;
import edu.princeton.cs.algs4.StdOut;

import java.util.Arrays;

class BinarySearch {
    public static int rank(int key, int[] a) {
        int lo = 0, hi = a.length - 1, mid;

        while (lo <= hi) {
            mid = lo + (hi - lo) / 2;
            if (key < a[mid]) hi = mid - 1;
            else if (key > a[mid]) lo = mid + 1;
            else return mid;
        }
        return -1;
    }

    public static int rrank(int key, int[] a, int low, int high) {
        if (low > high) return -1;

        int mid = low + (high - low) / 2;
        if (key < a[mid])
            return rrank(key, a, low, mid - 1);
        else if (key > a[mid])
            return rrank(key, a, mid + 1, high);
        else
            return mid;
    }

    public static void main(String[] args) {
        int[] array={1,2,3,4,5,6,7,8,9,10};

        StdOut.println("5's position: "+rank(5,array));
        StdOut.println("5's position: "+rrank(5,array,0,array.length-1));
    }
}

File: Algs4/src/1.Basic/C/BinarySearch.c
#include <stdio.h>

int BinarySearch(int key,int arr[],int n){
    int low=0,high=n-1,mid;

    while(low<=high){
        mid=low+(high-low)/2;
        if(key<arr[mid]) high=mid-1;
        else if(key>arr[mid]) low=mid+1;
        else return mid;
    }
    return -1;
}

int rBSearch(int key,int arr[],int low,int high){
    if(low>high) return -1;

    int mid=low+(high-low)/2;
    if(key<arr[mid])
        return rBSearch(key,arr,low,mid-1);
    else if(key>arr[mid]) 
        return rBSearch(key,arr,mid+1,high);
    return mid;
}

int rBinarySearch(int key,int arr[],int n){
    return rBSearch(key,arr,0,n-1);
}

int main(void)
{
    int arr[10]={0,1,2,3,4,5,6,7,8,9};

    printf("5's position: %d\n",BinarySearch(5,arr,10));
    printf("5's position: %d\n",rBinarySearch(5,arr,10));
    return 0;
}